aesd-char-driver/main.c

```diff
@@ -0,0 +1,257 @@
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+#include "linux/slab.h"
+#include "linux/string.h"
+
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Anuhya Kuraparthy"); 
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev;
+    PDEBUG("open");
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    filp->private_data = NULL;
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *dev;
+    struct aesd_buffer_entry *tmp_buf;
+    int tmp_buf_count = 0;
+    size_t offset_bytes;
+    dev = filp->private_data;
+    
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    
+    mutex_lock(&aesd_device.lock);
+    tmp_buf = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->buffer, *f_pos, &offset_bytes);
+    
+    if( tmp_buf == NULL )
+    {
+       // *f_pos = 0;
+        goto handle_error;
+    }
+    
+    if( (tmp_buf->size - offset_bytes) < count )
+    {
+        *f_pos += tmp_buf->size - offset_bytes;
+        tmp_buf_count = tmp_buf->size - offset_bytes;
+    }
+    else
+    {
+        *f_pos += count;
+        tmp_buf_count = count;
+    }
+    
+    if( copy_to_user(buf, tmp_buf->buffptr+offset_bytes, tmp_buf_count))
+    {
+        retval = -EFAULT;
+        goto handle_error;
+    }
+    
+    retval = tmp_buf_count;
+
+    handle_error : mutex_unlock(&aesd_device.lock);
+    
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    char *tmp_buf;		//tmp_buffer
+    bool packet_flag = false;	//packet_send
+    struct aesd_dev *dev;	//dev
+    int packet_len = 0;	//tmp_store
+    struct aesd_buffer_entry write_entry;	//write_buffer
+    char *ret;			//replaced_buffer
+    int i;
+    ssize_t retval = 0;	//?
+    int extra;			//tmp_total_size
+    
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    
+    dev = filp->private_data;
+    mutex_lock(&aesd_device.lock);
+    
+    tmp_buf = (char *)kmalloc(count, GFP_KERNEL);
+    if( tmp_buf == NULL )
+    {
+        retval = -ENOMEM;
+        goto end;
+    }
+    
+    if(copy_from_user(tmp_buf, buf, count))
+    {
+        retval = -EFAULT;
+        goto free_memory;
+    }
+    
+    for(i=0; i< count; i++)
+    {
+        if(tmp_buf[i] == '\n')
+        {
+            packet_flag = true;
+            packet_len = i+1;
+            break;
+        }
+    }
```

Comment:
@anuh7 do the lines through here need to be in the locking interval?  See discussion from class last week.